File: session-3/page-objects-checkout/node/test/pages/catalog.page.js
// CatalogPage — the product catalog/grid reached from the home screen.
// Owns the "Shop All" entry, the search box, the result grid, and the grid's
// cart icon. Tests call open / search / openFirstResult / openCart in plain language.
//
// Locators carried over verified from the Session 2 search->checkout capstone.

class CatalogPage {
  // --- locators: the single source of truth for this screen ---
  get shopAll() {
    // Entry point on the home screen that opens the full catalog.
    return $('~Shop All')
  }
  get searchField() {
    // The catalog's only EditText.
    return $('android=new UiSelector().className("android.widget.EditText")')
  }
  get firstResult() {
    // Product cards are clickable images; the first one is the first search hit.
    return $('android=new UiSelector().className("android.widget.ImageView").clickable(true).instance(0)')
  }
  get cartIcon() {
    // Grid app-bar buttons in order: [back/menu](0) · [sort](1) · [cart](2).
    return $('android=new UiSelector().className("android.widget.Button").instance(2)')
  }

  /** From the home screen, open the catalog and wait until search is interactive. */
  async open() {
    await this.shopAll.waitForDisplayed({ timeout: 20000 })
    await this.shopAll.click()
    await this.searchField.waitForDisplayed({ timeout: 20000 })
  }

  /** Type a query. Focus first — the grid filters reactively as you type. */
  async search(query) {
    await this.searchField.click()
    await this.searchField.clearValue()
    await this.searchField.addValue(query)
    try { await driver.hideKeyboard() } catch { /* keyboard already down */ }
  }

  /** Open the first search result's detail screen. */
  async openFirstResult() {
    await this.firstResult.waitForDisplayed({ timeout: 20000 })
    await this.firstResult.click()
  }

  /** Tap the grid's cart icon to open the cart. */
  async openCart() {
    await this.cartIcon.waitForDisplayed({ timeout: 20000 })
    await this.cartIcon.click()
  }
}

export default new CatalogPage()

File: session-3/page-objects-checkout/python/pages/catalog_page.py
"""CatalogPage — the product catalog/grid reached from the home screen.

Owns the "Shop All" entry, the search box, the result grid, and the grid's
cart icon.

Locators carried over verified from the Session 2 search->checkout capstone.
"""

from appium.webdriver.common.appiumby import AppiumBy
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC


class CatalogPage:
    # --- locators: the single source of truth for this screen ---
    SHOP_ALL = (AppiumBy.ACCESSIBILITY_ID, "Shop All")
    SEARCH = (AppiumBy.ANDROID_UIAUTOMATOR, 'new UiSelector().className("android.widget.EditText")')
    # Product cards are clickable images; the first one is the first search hit.
    FIRST_RESULT = (AppiumBy.ANDROID_UIAUTOMATOR, 'new UiSelector().className("android.widget.ImageView").clickable(true).instance(0)')
    # Grid app-bar buttons in order: [back/menu](0) · [sort](1) · [cart](2).
    CART_ICON = (AppiumBy.ANDROID_UIAUTOMATOR, 'new UiSelector().className("android.widget.Button").instance(2)')

    def __init__(self, driver):
        self.driver = driver

    def open(self):
        """From the home screen, open the catalog and wait until search is ready."""
        WebDriverWait(self.driver, 20).until(EC.element_to_be_clickable(self.SHOP_ALL)).click()
        WebDriverWait(self.driver, 20).until(EC.visibility_of_element_located(self.SEARCH))

    def search(self, query):
        """Type a query. Focus first — the grid filters reactively as you type."""
        field = WebDriverWait(self.driver, 20).until(EC.element_to_be_clickable(self.SEARCH))
        field.click()
        field.clear()
        field.send_keys(query)
        try:
            self.driver.hide_keyboard()
        except Exception:
            pass  # keyboard already down

    def open_first_result(self):
        """Open the first search result's detail screen."""
        WebDriverWait(self.driver, 20).until(EC.element_to_be_clickable(self.FIRST_RESULT)).click()

    def open_cart(self):
        """Tap the grid's cart icon to open the cart."""
        WebDriverWait(self.driver, 20).until(EC.element_to_be_clickable(self.CART_ICON)).click()
